### DISCUSSION TITLE
Why is alembic altering columns that remain unchanged?

### DISCUSSION THREAD

**Discussion by maham360** [sqlalchemy/alembic#1023] [Usage Questions]
I have this code in my migration file:

```
def upgrade():
    # ### commands auto generated by Alembic - please adjust! ###
    op.alter_column('notification_history', 'international',
               existing_type=sa.BOOLEAN(),
               nullable=False)
    op.alter_column('notifications', 'international',
               existing_type=sa.BOOLEAN(),
               nullable=False)
    op.alter_column('service_callback', 'bearer_token',
               existing_type=sa.VARCHAR(),
               nullable=False)
```
And here are those columns in my python model:

```
international = db.Column(db.Boolean, nullable=False, default=False)

international = db.Column(db.Boolean, nullable=False, default=False)

_bearer_token = db.Column("bearer_token", db.String(), nullable=False)
```
The datatype "boolean" is the same in the model and database so I'm not sure why alembic is stating the column is being altered? I'm using Flask-SQLAlchemy

**Comment by zzzeek** [comment]
those alters are changing the nullability of the columns to NOT NULL.   If they are NULL in the database schema, that's why these changes would be produced.  note there's no change in column type being proposed (it's setting "existing_type" , which other databases that are not PostgreSQL need to know when altering other aspects of a column).

**Reply by maham360** [reply]
Thank you that correct my issue! Also I just noticed this table 'service_callback':

```
    op.alter_column('service_callback', 'bearer_token',
               existing_type=sa.VARCHAR(),
               nullable=False)
```
Doesn't exist in my model but it does in my database. So here Alembic is saying my database has the table set to 'nullable=False'?

**Reply by zzzeek** [reply]
if `service_callback` is not in your model, then Alembic would try to say that the whole `service_callback` table should be dropped.  So your model may have this table represented somehow.

**Reply by maham360** [reply]
Yes I see. So my table `services_history` does not exist in my model but the table `services` does. Both tables have `prefix_sms` as a column. I thought if I correct the `services` table, it will correct `services_history`. 

`    op.alter_column('services_history', 'prefix_sms',
               existing_type=sa.BOOLEAN(),
               nullable=False)`

The issue I am facing is `prefix_sms` is not nullable, so in my model I changed it to True, and the migration file is showing the change under the `services` table as:

`    op.alter_column('services', 'prefix_sms',
               existing_type=sa.BOOLEAN(),
               nullable=True)`

I don't know how to correct it from here.

**Reply by zzzeek** [reply]
> Yes I see. So my table `services_history` does not exist in my model but the table `services` does. Both tables have `prefix_sms` as a column. I thought if I correct the `services` table, it will correct `services_history`.

alembic deals with each table individually, so you would address each table on a case by case basis.



> 
> ` op.alter_column('services_history', 'prefix_sms', existing_type=sa.BOOLEAN(), nullable=False)`
> 
> The issue I am facing is `prefix_sms` is not nullable, so in my model I changed it to True, and the migration file is showing the change under the `services` table as:

"not nullable" in the database, if you wanted to sync your model with the database, means you'd have nullable=False in your model also.  otherwise not sure what outcome you are looking for.
